src/main/java/com/simibubi/create/content/trains/entity/Train.java

```diff
@@ -1090,6 +1146,66 @@ public Couple<Couple<TrackNode>> getEndpointEdges() {
 			.map(tp -> Couple.create(tp.node1, tp.node2));
 	}
 
+
+	/**
+	 * Collision detection cache structure.
+	 * Stores precomputed train segment positions to avoid repeated graph lookups.
+	 */
+	private static class CollisionCache {
+		int segmentCount;
+
+		// Position arrays - precomputed to avoid repeated graph lookups
+		Vec3[] start;
+		Vec3[] end;
+
+		// Direction vectors (normalized) - precomputed for intersection tests
+		Vec3[] direction;
+		double[] segmentLength;
+
+		// Metadata
+		ResourceKey<Level> dimension;
+		boolean valid;
+
+		CollisionCache(int maxSegments) {
+			this.segmentCount = 0;
+			this.start = new Vec3[maxSegments];
+			this.end = new Vec3[maxSegments];
+			this.direction = new Vec3[maxSegments];
+			this.segmentLength = new double[maxSegments];
+			this.valid = false;
+		}
+
+		void invalidate() {
+			this.valid = false;
+		}
+
+		boolean isValid() {
+			return valid && segmentCount > 0;
+		}
+
+		/**
+		 * Add a segment to the cache with precomputed direction and length
+		 */
+		void addSegment(Vec3 startPos, Vec3 endPos, int index) {
+			if (index >= start.length)
```

Comment:
I don't like this check, it could silently hide a critical bug, and since this is all internal-use we know exactly how big the arrays are and should be able to confidently insert elements at all valid indices

Reply:
We should not be generating segments past the maximum expected index. Because of this check it will be extremely difficult to tell if that is happening, but if we remove it it'll just crash in dev and be obvious that some other invariant is not being upheld.

I'm a fan of range checks for cases where you cannot trust the inputs to your function, but this is self-contained and should need no such validation